lib/nerve.rb

```diff
@@ -45,15 +43,41 @@ def run
       end
 
       begin
-        sleep
-      rescue StandardError => e
+        loop do
+          # Check that watcher threads are still alive, auto-remediate if they
+          # are not. Sometimes zookeeper flakes out or connections are lost to
+          # remote datacenter zookeeper clusters, failing is not an option
+          relaunch = []
+          @watchers.each do |name, watcher_thread|
+            unless watcher_thread.alive?
+              log.warn "nerve: watcher #{name} not alive, attempting to relaunch"
+              relaunch << name
+            end
+          end
+
+          relaunch.each do |name|
+            begin
+              reap_watcher(name)
+            rescue => e
+              log.warn "nerve: could not reap #{name}, got #{e.inspect}"
+            end
+            launch_watcher(name, @services[name])
+          end
+
+          unless @heartbeat_path.nil?
+            FileUtils.touch(@heartbeat_path)
+          end
+
+          sleep 10
+        end
+      rescue => e
         log.error "nerve: encountered unexpected exception #{e.inspect} in main thread"
         raise e
       ensure
         $EXIT = true
         log.warn 'nerve: reaping all watchers'
         @watchers.each do |name, watcher_thread|
-          reap_watcher(name)
+          reap_watcher(name) rescue 'passing'
```

Comment:
maybe something like `rescue "watcher #{name} could not be immediately reaped; skipping"`